main/src/library/Iterator.flix

```diff
@@ -589,199 +524,184 @@ namespace Iterator {
-    ///
-    @Lazy
-    pub def intercalate(iterA: Iterator[a, r1], iterB: Iterator[Iterator[a, r2], r3]): Iterator[a, r1 and r2 and r3] \ Write(r1, r3) =
-        let Iterator(doneA, nextA) = iterA;
-        let Iterator(doneB, nextB) = iterB;
-        let r1 = Scoped.regionOf(iterA);
-        let xs = ref Nil @ r1;
-        let r3 = Scoped.regionOf(iterB);
-        let isReversed = ref false @ r1;
-        let nextAA = () -> {
-            let x = nextA();
-            xs := x :: deref xs;
-            x
-        };
-        let doneAA = () -> {
-            if (doneA()) {
-                if (deref isReversed)
-                    ()
-                else {
-                    xs := List.reverse(deref xs);
-                    isReversed := true
-                };
-                true
+    /// The original iterators `sep` and `iter` should not be reused.
+    ///
+    pub def intercalate(sep: Iterator[a, ef2, r], iter: Iterator[Iterator[a, ef1, r], ef1, r]): Iterator[a, ef1 and ef2 and r, r] \ Write(r) =
+        let Iterator(rc, _) = iter;
+        let start = ref true @ rc;
+        let sepList = ref Nil @ rc;
+        let step = innerIter -> {
+            // Cache `sepList` first time in the loop. Unfortunately we must test for Nil inside the loop [test each time]
+            // to prevent the effect `ef2` leaking to the top level of the function.
+            match (deref sepList) {
+                case Nil => sepList := toList(sep)
+                case _   => ()
+            };
+            if (deref start) {
+                start := false;
+                innerIter
+            } else {
+                append(ofList(rc, deref sepList), innerIter)
             }
-            else
-                false
         };
-        let cursor = ref None @ r3;
-        let mkCursor = i -> { // Appends an iterator with the contents of `iterA`. Boolean formula: r1 and formula of i
-            if (deref isReversed)
-                // `iterA` has been consumed, so construct a new one from `xs`
-                cursor := Some(append(i, List.iterator(r1, deref xs)))
-            else
-                cursor := Some(append(i, Iterator(doneAA, nextAA)))
-        };
-        let done1 = () -> match deref cursor {
-            case None    => match forward(match Iterator(d, _) -> not d(), doneB, nextB) {
-                case None    => true
-                case Some(i) =>
-                    if (doneB())
-                        cursor := Some(append(i, new Iterator(r1))) // Has Boolean formula r2 since it is the last iterator of `iterB`.
-                    else                                                     // Has to be appended with an empty iterator to satisfy formula `r1 and formula of i` in `mkCursor`
-                        mkCursor(i);                                         // Same reason for appending with empty iterator below
-                    false
+        flatMap(step, iter)
+
+    ///
+    /// Helper for append.
```

Comment:
I don't think append uses this?

Reply:
Used by `intercalate` not `append`` - fixed